dotnet/src/Connectors/Connectors.Memory.Redis/RedisVectorRecordStore.cs

```diff
@@ -0,0 +1,385 @@
+        }
+        else
+        {
+            properties = VectorStoreRecordPropertyReader.FindProperties(typeof(TRecord), supportsMultipleVectors: true);
+        }
+
+        // Validate property types and store for later use.
+        VectorStoreRecordPropertyReader.VerifyPropertyTypes([properties.keyProperty], s_supportedKeyTypes, "Key");
+        VectorStoreRecordPropertyReader.VerifyPropertyTypes(properties.vectorProperties, s_supportedVectorTypes, "Vector");
+
+        this._keyPropertyInfo = properties.keyProperty;
+        this._keyJsonPropertyName = VectorStoreRecordPropertyReader.GetSerializedPropertyName(this._keyPropertyInfo);
+
+        this._dataPropertyNames = properties
+            .dataProperties
+            .Select(VectorStoreRecordPropertyReader.GetSerializedPropertyName)
+            .ToArray();
+
+        this._dataAndVectorPropertyNames = this._dataPropertyNames
+            .Concat(properties.vectorProperties.Select(VectorStoreRecordPropertyReader.GetSerializedPropertyName))
+            .ToArray();
+
+        // Assign Mapper.
+        if (this._options.MapperType == RedisRecordMapperType.JsonNodeCustomMapper)
+        {
+            if (this._options.JsonNodeCustomMapper is null)
+            {
+                throw new ArgumentException($"The {nameof(RedisVectorRecordStoreOptions<TRecord>.JsonNodeCustomMapper)} option needs to be set if a {nameof(RedisVectorRecordStoreOptions<TRecord>.MapperType)} of {nameof(RedisRecordMapperType.JsonNodeCustomMapper)} has been chosen.", nameof(options));
+            }
+
+            this._mapper = this._options.JsonNodeCustomMapper;
+        }
+        else
+        {
+            this._mapper = new RedisVectorStoreRecordMapper<TRecord>(this._keyJsonPropertyName);
+        }
+    }
+
+    /// <inheritdoc />
+    public async Task<TRecord> GetAsync(string key, GetRecordOptions? options = null, CancellationToken cancellationToken = default)
+    {
+        Verify.NotNullOrWhiteSpace(key);
+
+        // Create Options
+        var collectionName = this.ChooseCollectionName(options?.CollectionName);
+        var maybePrefixedKey = this.PrefixKeyIfNeeded(key, collectionName);
+
+        // Get the redis value.
+        var redisResult = await RunOperationAsync(
+            collectionName,
+            "GET",
+            () => options?.IncludeVectors is true ?
+                this._database
+                    .JSON()
+                    .GetAsync(maybePrefixedKey) :
+                this._database
+                    .JSON()
+                    .GetAsync(maybePrefixedKey, this._dataPropertyNames)).ConfigureAwait(false);
+
+        // Check if the key was found before trying to serialize the result.
+        if (redisResult.IsNull)
+        {
+            throw new VectorStoreOperationException($"Could not find document with key '{key}'");
```

Comment:
Is our strategy to throw an exception when document is not found for all connectors? I'm wondering if we can return `null`? From usage perspective, it will be easier to handle `null` value comparing to wrapping `GetAsync` method in `try/catch` block. Same for batch scenario and `IAsyncEnumerable` result.

Reply:
Some database SDKs throw by default when trying to get an item that doesn't exist, so in that case we'd have to be able to catch the right type of exception and return null.  I'm concerned that those exceptions might not always be deterministic using code or we have to look at the exception message to determine what the exception means.
I was actually returning null to begin with, including for batch results, but it became really messy really fast, since you then have to leave null gaps in the resultset, and the resultset has to be in the same order as the requests to determine which ones were not retrieved.  Also, not all databases may return you partial results for batch operations, in which case we don't have the ability to do that ourselves either.

Reply:
On the other hand, it's better to keep it consistent for result type, because otherwise if one connector throws an exception and other one returns `null`, it won't be possible to use the same code with abstractions and easily switch between different connectors, because the result should be handled differently. In this case, throwing an exception in such cases is a good approach, which can be used in all connectors.